Added basic functionality for url encoding

ben-jones opened the pull request:
Created a urlEncode module with a variety of methods for url encoding and a series of unit tests to validate operation. Currently, cookie encoding does not work, but url encoding does.

ben-jones committed:
Added basic functionality for url encoding
Created a urlEncode module with a variety of methods for url encoding and a series of unit tests to validate operation

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+def decode(data):
+  """Decode the given data after matching the url hiding format
+
+  Parameters: data- the url and cookies to be decoded in the form of a
```
This is mildly confusing. In the previous functions, data is just a string but now data is a dict in this function. Maybe rename data?

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+def pickDomain():
+  """Pick a random domain from the list"""
+  domains = ['live.com', 'microsoft.com', 'baidu.com', 'hao123.com']
+  return domains[randint(0, len(domains)-1)]
```
``` python
from random import choice
return choice(domains)
```

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+
+  cookies = []
+  while data != '':
+      encodeAsCookie(data[:BYTES_PER_COOKIE])
```
check if data has enough bytes first, otherwise it might lead to a `IndexError: string index out of range`

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+  matches = re.search(pattern, url)
+  data = matches.group('hash')
+  # strip any padding
+  data = data.rstrip('ABCDEF')
```
what about padding on the left side? `data.strip` -- I'm not if there will be padding on the left side, but just making sure.

ben-jones replied:
There should only be padding on the right, but I changed it anyway.

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+  """
+  url = data['url']
+  cookies = data['cookie']
+  retVal = ''
```
It's more efficient to have retVal as a list than a string. Strings are immutable in python, so every time you append data to it, it gets created again. Just do `retVal.append(decodeAsMarket(url))` and in the end do `return ''.join(retVal)`. Also, maybe pick a more expressive variable name than retVal?

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,140 @@
+
+  """
+  if encodingType not in AVAILABLE_TYPES:
+      raise(UrlEncodeError("Bad encoding type. Please refer to url-encode.AVAILABLE_TYPES for available options"))
```
Line wrap please

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+    while index < 80:
+      stringy += characters[randint(0,15)]
+      index += 1
+    return stringy
```
lol stringy? :P

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+    #include a test to be sure that we can correctly decode padding
+    testData[4] += "ABCD"
+    for datum in testData:
+      url = 'click.live.com' + '?qs=' + datum
```
yeah, don't do this. Check out http://docs.python.org/2/library/urllib.html#examples

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+
+    testData = [ self.gen_random_hex_chars() for index in range(5)]
+    testData[4] = testData[4][:76]
+    testStrings = [binascii.unhexlify(testData[index]) for index in range(5)]
```
why not do -

```
[binascii.unhexlify(datum) for datum in testData]
```

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+    """Verify that decodeAsMarket correctly decodes data"""
+
+    testData = [ self.gen_random_hex_chars() for index in range(5)]
+    testData[4] = testData[4][:76]
```
Why 76? Sounds like a magic number ;)

ben-jones replied:
This was a just a random number that I picked to perform this test.

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+    testData = ['some text', 'more text', 
+                'even more text that seems moderately lon']
+    for datum in testData:
+        output = urlEncode.encodeAsMarket(datum)
```
output? what is output? Maybe use decodedUrl?

ben-jones committed:
Updated code after initial review
Updated code so that we are doing line wrapping where appropriate, added some mroe error checking, and added some optimizations

gsathya commented on code/verifyUrlEncode.py:
```diff
@@ -0,0 +1,112 @@
+                  '8','9','a','b','c','d','e','f']
+    index = 0
+    stringy = ''
+    while index < 80:
```
Why use a while?

```
random_hex = []
for x in range(80):
  random_hex.append(choice(characters))
return ''.join(random_hex)
```

gsathya commented on code/urlEncode.py:
```diff
@@ -0,0 +1,145 @@
+def pickDomain():
+  """Pick a random domain from the list"""
+  domains = ['live.com', 'microsoft.com', 'baidu.com', 'hao123.com']
+  return choice(domains)
```
use `choice` instead of randint

ben-jones committed:
Added support for cookies
Implemented functionality to hide data in cookies and updated unit tests to appropriately cover the functions.

ben-jones commented:
I chose not to inline the functionality of the test for 2 reasons. 1) I wanted to be sure that I was the test was not dependent on another function and 2) I wanted to add flexibility to the unittest. If we suddenly think of an edge case, it is trivial to add it now. Ultimately, this is just personal preference though.

gsathya commented:
> I chose not to inline the functionality of the test for 2 reasons. 1) I wanted to be sure that I was the test was not dependent on another function and 2) I wanted to add flexibility to the unittest. If we suddenly think of an edge case, it is trivial to add it now. Ultimately, this is just personal preference though.

No no, in `self.assertEqual(urlEncode.decodeAsCookie(cookie), datum)`, you're deriving the value for `cookie` exactly as how you would in encodeAsCookie. Shouldn't you just hardcode the value of `cookie`? Like for ex  `self.assertEqual(urlEncode.decodeAsCookie("some_cookie_value_that_is_hardcoded", datum)`

ben-jones committed:
Responses to second round of code commits
Made changes in response to all comments except for the comments about url parsing. In particular, it is not possible to address comments about urlparse because it changes all hex characters to uppercase and we need uppercase to distinguish padding from actual data

gsathya commented:
2dbf1bc990ecb8076ad1938c88885f43a1002875 needs the second line in the commit message to be line wrapped. I suggest using M-x fill-paragraph.

Amend the commit message and force push it. I'll merge it.

gsathya commented:
> Amend the commit message and force push it. I'll merge it.

I've fixed it.

gsathya commented:
Merged. Thanks!
